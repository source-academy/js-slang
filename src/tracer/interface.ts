import { StepperExpression, StepperPattern } from './nodes'

export interface StepperBaseNode {
  type: string
  isContractible(): boolean
  isOneStepPossible(): boolean
  contract(): StepperBaseNode
  oneStep(): StepperBaseNode
  substitute(id: StepperPattern, value: StepperExpression): StepperBaseNode
  freeNames(): string[]
  allNames(): string[]
  rename(before: string, after: string): StepperBaseNode
}
