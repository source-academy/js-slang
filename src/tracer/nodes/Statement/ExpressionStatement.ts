import { Comment, SourceLocation, ExpressionStatement } from 'estree'
import { StepperBaseNode } from '../../interface'
import { convert } from '../../generator'
import { StepperExpression, StepperPattern } from '..'
import { redex } from '../..'

export class StepperExpressionStatement implements ExpressionStatement, StepperBaseNode {
  type: 'ExpressionStatement'
  expression: StepperExpression
  leadingComments?: Comment[] | undefined
  trailingComments?: Comment[] | undefined
  loc?: SourceLocation | null | undefined
  range?: [number, number] | undefined

  constructor(
    expression: StepperExpression,
    leadingComments?: Comment[] | undefined,
    trailingComments?: Comment[] | undefined,
    loc?: SourceLocation | null | undefined,
    range?: [number, number] | undefined
  ) {
    this.type = 'ExpressionStatement'
    this.expression = expression
    this.leadingComments = leadingComments
    this.trailingComments = trailingComments
    this.loc = loc
    this.range = range
  }

  static create(node: ExpressionStatement) {
    return new StepperExpressionStatement(
      convert(node.expression) as StepperExpression,
      node.leadingComments,
      node.trailingComments,
      node.loc,
      node.range
    )
  }

  isContractible(): boolean {
    return this.expression.isContractible()
  }
  isOneStepPossible(): boolean {
    return this.expression.isOneStepPossible()
  }
  contract(): StepperExpressionStatement {
    return new StepperExpressionStatement(
      this.expression.oneStep(),
      this.leadingComments,
      this.trailingComments,
      this.loc,
      this.range
    )
  }

  contractEmpty() {
    // Handle cases such as 1; 2; -> 2;
    redex.preRedex = [this]
    redex.postRedex = []
  }

  oneStep(): StepperExpressionStatement {
    return new StepperExpressionStatement(
      this.expression.oneStep(),
      this.leadingComments,
      this.trailingComments,
      this.loc,
      this.range
    )
  }

  substitute(id: StepperPattern, value: StepperExpression): StepperBaseNode {
    return new StepperExpressionStatement(
      this.expression.substitute(id, value),
      this.leadingComments,
      this.trailingComments,
      this.loc,
      this.range
    )
  }

  freeNames(): string[] {
    return this.expression.freeNames()
  }

  allNames(): string[] {
    return this.expression.allNames()
  }

  rename(before: string, after: string): StepperExpressionStatement {
    return new StepperExpressionStatement(
      this.expression.rename(before, after),
      this.leadingComments,
      this.trailingComments,
      this.loc,
      this.range
    )
  }
}
