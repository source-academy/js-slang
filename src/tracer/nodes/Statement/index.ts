import { StepperBlockStatement } from './BlockStatement'
import { StepperExpressionStatement } from './ExpressionStatement'
import { StepperFunctionDeclaration } from './FunctionDeclaration'
import { StepperReturnStatement } from './ReturnStatement'
import { StepperVariableDeclaration } from './VariableDeclaration'

export type StepperStatement =
  | StepperExpressionStatement
  | StepperVariableDeclaration
  | StepperBlockStatement
  | StepperReturnStatement
  | StepperFunctionDeclaration
