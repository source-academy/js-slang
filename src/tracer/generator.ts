/*
TODO: Write docs  
convert estree into corresponding stepper type
Every class should have the following properties
- basic StepperBaseNodeInterface
- constructor: create new AST from class type StepperBaseNode
- static create: factory method to parse estree to StepperAST
*/

import * as es from 'estree'
import { generate } from 'astring'
import { StepperBinaryExpression } from './nodes/Expression/BinaryExpression'
import { StepperUnaryExpression } from './nodes/Expression/UnaryExpression'
import { StepperLiteral } from './nodes/Expression/Literal'
import { StepperBaseNode } from './interface'
import { StepperExpressionStatement } from './nodes/Statement/ExpressionStatement'
import { StepperProgram } from './nodes/Program'
import {
  StepperVariableDeclaration,
  StepperVariableDeclarator
} from './nodes/Statement/VariableDeclaration'
import { StepperIdentifier } from './nodes/Expression/Identifier'
import { StepperBlockStatement } from './nodes/Statement/BlockStatement'
import { StepperIfStatement } from './nodes/Statement/IfStatement'
import { StepperConditionalExpression } from './nodes/Expression/ConditionalExpression'
import { StepperArrowFunctionExpression } from './nodes/Expression/ArrowFunctionExpression'
import { StepperFunctionApplication } from './nodes/Expression/FunctionApplication'
import { StepperReturnStatement } from './nodes/Statement/ReturnStatement'
import { StepperFunctionDeclaration } from './nodes/Statement/FunctionDeclaration'
import { StepperArrayExpression } from './nodes/Expression/ArrayExpression'
import { StepperLogicalExpression } from './nodes/Expression/LogicalExpression'
import { StepperBlockExpression } from './nodes/Expression/BlockExpression'
import { isBuiltinFunction } from './builtins'
const undefinedNode = new StepperLiteral('undefined')

const nodeConverters: { [Key: string]: (node: any) => StepperBaseNode } = {
  Literal: (node: es.SimpleLiteral) => StepperLiteral.create(node),
  UnaryExpression: (node: es.UnaryExpression) => StepperUnaryExpression.create(node),
  BinaryExpression: (node: es.BinaryExpression) => StepperBinaryExpression.create(node),
  LogicalExpression: (node: es.LogicalExpression) => StepperLogicalExpression.create(node),
  FunctionDeclaration: (node: es.FunctionDeclaration) => StepperFunctionDeclaration.create(node),
  ExpressionStatement: (node: es.ExpressionStatement) => StepperExpressionStatement.create(node),
  ConditionalExpression: (node: es.ConditionalExpression) =>
    StepperConditionalExpression.create(node),
  ArrowFunctionExpression: (node: es.ArrowFunctionExpression) =>
    StepperArrowFunctionExpression.create(node),
  ArrayExpression: (node: es.ArrayExpression) => StepperArrayExpression.create(node),
  CallExpression: (node: es.CallExpression) =>
    StepperFunctionApplication.create(node as es.SimpleCallExpression),
  ReturnStatement: (node: es.ReturnStatement) => StepperReturnStatement.create(node),
  Program: (node: es.Program) => StepperProgram.create(node),
  VariableDeclaration: (node: es.VariableDeclaration) => StepperVariableDeclaration.create(node),
  VariableDeclarator: (node: es.VariableDeclarator) => StepperVariableDeclarator.create(node),
  Identifier: (node: es.Identifier) => {
    if (node.name === 'NaN') {
      return new StepperLiteral(NaN, 'NaN')
    } else if (node.name === 'Infinity') {
      return new StepperLiteral(Infinity, 'Infinity')
    } else {
      return StepperIdentifier.create(node)
    }
  },
  BlockStatement: (node: es.BlockStatement) => StepperBlockStatement.create(node),
  IfStatement: (node: es.IfStatement) => StepperIfStatement.create(node)
}

export function convert(node: es.BaseNode): StepperBaseNode {
  const converter = nodeConverters[node.type as keyof typeof nodeConverters]
  return converter ? converter(node as any) : undefinedNode
}

// Explanation generator
export function explain(redex: StepperBaseNode): string {
  const explainers = {
    UnaryExpression: (node: StepperUnaryExpression) => {
      if (node.operator === '-') {
        return (
          'Unary expression evaluated, value ' +
          JSON.stringify((node.argument as StepperLiteral).value) +
          ' negated.'
        )
      } else if (node.operator === '!') {
        return (
          'Unary expression evaluated, boolean ' +
          JSON.stringify((node.argument as StepperLiteral).value) +
          ' negated.'
        )
      } else {
        throw new Error('Unsupported unary operator ' + node.operator)
      }
    },
    BinaryExpression: (node: StepperBinaryExpression) => {
      return 'Binary expression ' + generate(node) + ' evaluated'
    },
    LogicalExpression: (node: StepperLogicalExpression) => {
      if (node.operator == '&&') {
        return (node.left as StepperLiteral).value === false
          ? 'AND operation evaluated, left of operator is true, continue evaluating right of operator'
          : 'AND operation evaluated, left of operator is false, stop evaluation'
      } else if (node.operator == '||') {
        return (node.left as StepperLiteral).value === true
          ? 'OR operation evaluated, left of operator is true, stop evaluation'
          : 'OR operation evaluated, left of operator is false, continue evaluating right of operator'
      } else {
        throw new Error('Invalid operator')
      }
    },
    VariableDeclaration: (node: StepperVariableDeclaration) => {
      if (node.kind === 'const') {
        return (
          'Constant ' +
          node.declarations.map(ast => ast.id.name).join(', ') +
          ' declared and substituted into the rest of block'
        )
      } else {
        return '...'
      }
    },
    ReturnStatement: (node: StepperReturnStatement) => {
      if (!node.argument) {
        throw new Error('return argument should not be empty')
      }
      return generate(node.argument!) + ' returned'
    },
    FunctionDeclaration: (node: StepperFunctionDeclaration) => {
      return `Function ${node.id.name} declared, parameter(s) ${node.params.map(x =>
        generate(x)
      )} required`
    },
    ExpressionStatement: (node: StepperExpressionStatement) => {
      return generate(node.expression) + ' finished evaluating'
    },
    BlockStatement: (node: StepperBlockStatement) => {
      if (node.body.length === 0) {
        return 'Empty block expression evaluated'
      }
      return generate(node.body[0]) + ' finished evaluating'
    },
    BlockExpression: (_node: StepperBlockExpression) => {
      throw new Error('Not implemented')
    },
    ConditionalExpression: (node: StepperConditionalExpression) => {
      const test = node.test // test should have typeof literal
      if (test.type !== 'Literal') {
        throw new Error('Invalid conditional contraction. `test` should be literal.')
      }
      const testStatus = (test as StepperLiteral).value
      if (typeof testStatus !== 'boolean') {
        throw new Error(
          'Invalid conditional contraction. `test` should be boolean, got ' +
            typeof testStatus +
            ' instead.'
        )
      }
      if (testStatus === true) {
        return 'Conditional expression evaluated, condition is true, consequent evaluated'
      } else {
        return 'Conditional expression evaluated, condition is false, alternate evaluated'
      }
    },
    CallExpression: (node: StepperFunctionApplication) => {
      if (node.callee.type !== 'ArrowFunctionExpression' && node.callee.type !== 'Identifier') {
        throw new Error('`callee` should be function expression.')
      }

      // Determine whether the called function is built-in or not and create explanation accordingly
      const func: StepperArrowFunctionExpression = node.callee as StepperArrowFunctionExpression
      if (func.name && isBuiltinFunction(func.name)) {
        return `${func.name} runs`
        // @ts-ignore func.body.type can be StepperBlockExpression
      } else if (func.body.type === 'BlockStatement') {
        if (func.params.length === 0) {
          return '() => {...}' + ' runs'
        }
        const paramDisplay = func.params.map(x => x.name).join(', ')
        const argDisplay: string = node.arguments
          .map(x =>
            (x.type === 'ArrowFunctionExpression' || x.type === 'Identifier') &&
            x.name !== undefined
              ? x.name
              : generate(x)
          )
          .join(', ')
        return 'Function ' + func.name + ' takes in ' + argDisplay + ' as input ' + paramDisplay
      } else {
        if (func.params.length === 0) {
          return generate(func) + ' runs'
        }
        return (
          node.arguments.map(x => generate(x)).join(', ') +
          ' substituted into ' +
          func.params.map(x => x.name).join(', ') +
          ' of ' +
          generate(func)
        )
      }
    },
    ArrowFunctionExpression: (_node: StepperArrowFunctionExpression) => {
      throw new Error('Not implemented.')
    },
    IfStatement: (node: StepperIfStatement) => {
      if (node.test instanceof StepperLiteral) {
        if (node.test.value) {
          return 'If statement evaluated, condition true, proceed to if block'
        } else {
          return 'If statement evaluated, condition false, proceed to else block'
        }
      } else {
        throw new Error('Not implemented')
      }
    },
    Default: (_: StepperBaseNode) => {
      return '...'
    }
  }
  //@ts-ignore gracefully handle default ast node
  const explainer = explainers[redex.type] ?? explainers.Default
  return explainer(redex)
}
