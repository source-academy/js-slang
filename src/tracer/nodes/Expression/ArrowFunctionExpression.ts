import { ArrowFunctionExpression, Comment, SourceLocation } from 'estree'
import { StepperBaseNode } from '../../interface'
import { StepperExpression, StepperPattern } from '..'
import { convert } from '../../generator'
import { getFreshName } from '../../utils'
import { StepperBlockStatement } from '../Statement/BlockStatement'

export class StepperArrowFunctionExpression implements ArrowFunctionExpression, StepperBaseNode {
  type: 'ArrowFunctionExpression'
  params: StepperPattern[]
  body: StepperExpression
  expression: boolean
  generator: boolean
  async: boolean
  name?: string
  leadingComments?: Comment[]
  trailingComments?: Comment[]
  loc?: SourceLocation | null
  range?: [number, number]

  constructor(
    params: StepperPattern[],
    body: StepperExpression,
    name?: string,
    expression: boolean = true,
    generator: boolean = false,
    async: boolean = false,
    leadingComments?: Comment[],
    trailingComments?: Comment[],
    loc?: SourceLocation | null,
    range?: [number, number]
  ) {
    this.type = 'ArrowFunctionExpression'
    this.params = params
    this.expression = expression
    this.generator = generator
    this.async = async
    this.name = name
    this.leadingComments = leadingComments
    this.trailingComments = trailingComments
    this.loc = loc
    this.range = range
    this.body = body
  }

  static create(node: ArrowFunctionExpression) {
    return new StepperArrowFunctionExpression(
      node.params.map(param => convert(param) as StepperPattern),
      convert(node.body) as StepperExpression,
      undefined,
      node.expression,
      node.generator,
      node.async,
      node.leadingComments,
      node.trailingComments,
      node.loc,
      node.range
    )
  }

  isContractible(): boolean {
    return false
  }

  isOneStepPossible(): boolean {
    return false
  }

  contract(): StepperExpression {
    throw new Error('Cannot contract an arrow function expression')
  }

  oneStep(): StepperExpression {
    throw new Error('Cannot step an arrow function expression')
  }

  assignName(name: string): StepperArrowFunctionExpression {
    return new StepperArrowFunctionExpression(
      this.params,
      this.body,
      name,
      this.expression,
      this.generator,
      this.async,
      this.leadingComments,
      this.trailingComments,
      this.loc,
      this.range
    )
  }

  scanAllDeclarationNames(): string[] {
    const paramNames = this.params.map(param => param.name)

    let bodyDeclarations: string[] = []
    // @ts-expect-error gracefully handle block statement as block expression
    if (this.body.type === 'BlockStatement') {
      const body = this.body as StepperBlockStatement
      bodyDeclarations = body.body
        .filter(stmt => stmt.type === 'VariableDeclaration')
        .flatMap(decl => (decl as any).declarations.map((d: any) => d.id.name))
    }

    return [...paramNames, ...bodyDeclarations]
  }

  // TODO: Fix name handling for lambda
  substitute(
    id: StepperPattern,
    value: StepperExpression,
    upperBoundName?: string[]
  ): StepperExpression {
    const valueFreeNames = value.freeNames()
    const scopeNames = this.scanAllDeclarationNames()
    const repeatedNames = valueFreeNames.filter(name => scopeNames.includes(name))

    let protectedNamesSet = new Set([this.allNames(), upperBoundName ?? []].flat())
    repeatedNames.forEach(name => protectedNamesSet.delete(name))
    const protectedNames = Array.from(protectedNamesSet)
    const newNames = getFreshName(repeatedNames, protectedNames)
    const currentArrowFunction = newNames.reduce(
      (current: StepperArrowFunctionExpression, name: string, index: number) =>
        current.rename(repeatedNames[index], name) as StepperArrowFunctionExpression,
      this
    )
    if (currentArrowFunction.scanAllDeclarationNames().includes(id.name)) {
      return currentArrowFunction
    }

    return new StepperArrowFunctionExpression(
      currentArrowFunction.params,
      currentArrowFunction.body.substitute(
        id,
        value,
        currentArrowFunction.params.flatMap(p => p.allNames())
      ),
      currentArrowFunction.name,
      currentArrowFunction.expression,
      currentArrowFunction.generator,
      currentArrowFunction.async,
      currentArrowFunction.leadingComments,
      currentArrowFunction.trailingComments,
      currentArrowFunction.loc,
      currentArrowFunction.range
    )
  }

  freeNames(): string[] {
    const paramNames = this.params
      .filter(param => param.type === 'Identifier')
      .map(param => param.name)
    return this.body.freeNames().filter(name => !paramNames.includes(name))
  }

  allNames(): string[] {
    const paramNames = this.params
      .filter(param => param.type === 'Identifier')
      .map(param => param.name)
    return Array.from(new Set([paramNames, this.body.allNames()].flat()))
  }

  rename(before: string, after: string): StepperExpression {
    return new StepperArrowFunctionExpression(
      this.params.map(param => param.rename(before, after)),
      this.body.rename(before, after),
      this.name,
      this.expression,
      this.generator,
      this.async,
      this.leadingComments,
      this.trailingComments,
      this.loc,
      this.range
    )
  }
}
