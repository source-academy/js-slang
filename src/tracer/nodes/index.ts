import { StepperArrayExpression } from './Expression/ArrayExpression'
import { StepperArrowFunctionExpression } from './Expression/ArrowFunctionExpression'
import { StepperBinaryExpression } from './Expression/BinaryExpression'
import { StepperConditionalExpression } from './Expression/ConditionalExpression'
import { StepperFunctionApplication } from './Expression/FunctionApplication'
import { StepperIdentifier } from './Expression/Identifier'
import { StepperLiteral } from './Expression/Literal'
import { StepperLogicalExpression } from './Expression/LogicalExpression'
import { StepperUnaryExpression } from './Expression/UnaryExpression'

export type StepperExpression =
  | StepperUnaryExpression
  | StepperBinaryExpression
  | StepperLiteral
  | StepperPattern
  | StepperConditionalExpression
  | StepperFunctionApplication
  | StepperArrowFunctionExpression
  | StepperArrayExpression
  | StepperLogicalExpression
export type StepperPattern = StepperIdentifier

export const undefinedNode = new StepperLiteral('undefined', 'undefined')
