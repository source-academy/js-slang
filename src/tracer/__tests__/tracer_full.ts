import * as acorn from 'acorn'
import { getSteps } from '../steppers'
import { convert } from '../generator'
import * as astring from 'astring'
import { StepperBaseNode } from '../interface'
import { StepperProgram } from '../nodes/Program'
import { StepperExpressionStatement } from '../nodes/Statement/ExpressionStatement'
import { StepperArrowFunctionExpression } from '../nodes/Expression/ArrowFunctionExpression'

function codify(node: StepperBaseNode) {
  const steps = getSteps(convert(node))
  const stringify = (ast: StepperBaseNode) => {
    if (ast === undefined || ast!.type === undefined) {
      return ''
    }
    return astring.generate(ast)
  }
  return steps.map(prop => stringify(prop.ast).replace(/\n/g, '').replace(/\s+/g, ' '))
}

function acornParser(code: string): StepperBaseNode {
  return convert(acorn.parse(code, { ecmaVersion: 10 }))
}

describe('Expressions', () => {
  test('No extra step for UnaryExpression', async () => {
    const code = `
    - (1 + 2);
    `
    const steps = await codify(acornParser(code))
    expect(steps.join('')).toMatchSnapshot()
  }),
    test('Extra step for UnaryExpression', async () => {
      const code = `
    - (1 - 5);
    `
      const steps = await codify(acornParser(code))
      expect(steps.join('')).toMatchInlineSnapshot(`"-(1 - 5);-(1 - 5);--4;--4;4;4;"`)
    }),
    test('Unary and Binary Expressions', async () => {
      const code = `
    - 1 + 2 * 3 - (5 * 6 - 7);
    `
      const steps = await codify(acornParser(code))
      expect(steps.join('')).toMatchInlineSnapshot(
        `"-1 + 2 * 3 - (5 * 6 - 7);-1 + 2 * 3 - (5 * 6 - 7);-1 + 6 - (5 * 6 - 7);-1 + 6 - (5 * 6 - 7);5 - (5 * 6 - 7);5 - (5 * 6 - 7);5 - (30 - 7);5 - (30 - 7);5 - 23;5 - 23;-18;-18;"`
      )
    }),
    test('Logical Expression', async () => {
      const code = `
    !!!true || true;
    `
      const steps = await codify(acornParser(code))
      expect(steps.join('')).toMatchInlineSnapshot(
        `"!!!true || true;!!!true || true;!!false || true;!!false || true;!true || true;!true || true;false || true;false || true;true;true;"`
      )
    })
  test('Conditional Expression', async () => {
    const code = `
    (-1 * 3 === 3) ? 2 * 4 - 7 : 1 + 3 * 6;
    `
    const steps = await codify(acornParser(code))
    expect(steps.join('')).toMatchInlineSnapshot(
      `"-1 * 3 === 3 ? 2 * 4 - 7 : 1 + 3 * 6;-1 * 3 === 3 ? 2 * 4 - 7 : 1 + 3 * 6;-3 === 3 ? 2 * 4 - 7 : 1 + 3 * 6;-3 === 3 ? 2 * 4 - 7 : 1 + 3 * 6;false ? 2 * 4 - 7 : 1 + 3 * 6;false ? 2 * 4 - 7 : 1 + 3 * 6;1 + 3 * 6;1 + 3 * 6;1 + 18;1 + 18;19;19;"`
    )
  })
})

test('Test two statements', async () => {
  const code = `
    (1 + 2) * (3 + 4);
    3 * 5;
  `
  const steps = await codify(acornParser(code))
  expect(steps.join('\n')).toMatchInlineSnapshot(`
    "(1 + 2) * (3 + 4);3 * 5;
    (1 + 2) * (3 + 4);3 * 5;
    3 * (3 + 4);3 * 5;
    3 * (3 + 4);3 * 5;
    3 * 7;3 * 5;
    3 * 7;3 * 5;
    21;3 * 5;
    21;3 * 5;
    21;15;
    21;15;
    15;
    15;"
  `)
})

test('Test constant declaration substitution', async () => {
  const code = `
    const x = -1;
    x;
    const y = 2;
    y;
  `
  const steps = await codify(acornParser(code))
  expect(steps.join('\n')).toMatchInlineSnapshot(`
    "const x = -1;x;const y = 2;y;
    const x = -1;x;const y = 2;y;
    -1;const y = 2;y;
    -1;const y = 2;y;
    -1;2;
    -1;2;
    2;
    2;"
  `)
})

describe('Lambda expression', () => {
  test('Basic function', async () => {
    const code = `
      const y = 2;
      const f = x => x + y;
      f(1);
    `
    const steps = await codify(acornParser(code))
    expect(steps.join('\n')).toMatchInlineSnapshot(`
      "const y = 2;const f = x => x + y;f(1);
      const y = 2;const f = x => x + y;f(1);
      const f = x => x + 2;f(1);
      const f = x => x + 2;f(1);
      (x => x + 2)(1);
      (x => x + 2)(1);
      1 + 2;
      1 + 2;
      3;
      3;"
    `)
  }),
    test('Basic bi function', async () => {
      const code = `
      const add = (x, y) => x + y;
      add(2, 3);
    `
      const steps = await codify(acornParser(code))
      expect(steps.join('\n')).toMatchInlineSnapshot(`
        "const add = (x, y) => x + y;add(2, 3);
        const add = (x, y) => x + y;add(2, 3);
        ((x, y) => x + y)(2, 3);
        ((x, y) => x + y)(2, 3);
        2 + 3;
        2 + 3;
        5;
        5;"
      `)
    }),
    test('Currying', async () => {
      const code = `
      const add = x => y => x + y;
      add(2)(3);
    `
      const steps = await codify(acornParser(code))
      expect(steps.join('\n')).toMatchInlineSnapshot(`
        "const add = x => y => x + y;add(2)(3);
        const add = x => y => x + y;add(2)(3);
        (x => y => x + y)(2)(3);
        (x => y => x + y)(2)(3);
        (y => 2 + y)(3);
        (y => 2 + y)(3);
        2 + 3;
        2 + 3;
        5;
        5;"
      `)
    }),
    test('Recursive function call', async () => {
      const code = `
      const factorial = n => n === 0 ? 1 : n * factorial(n - 1);
      factorial(2);
    `
      const steps = await codify(acornParser(code))
      expect(steps.join('\n')).toMatchInlineSnapshot(`
        "const factorial = n => n === 0 ? 1 : n * factorial(n - 1);factorial(2);
        const factorial = n => n === 0 ? 1 : n * factorial(n - 1);factorial(2);
        (n => n === 0 ? 1 : n * factorial(n - 1))(2);
        (n => n === 0 ? 1 : n * factorial(n - 1))(2);
        2 === 0 ? 1 : 2 * (n => n === 0 ? 1 : n * factorial(n - 1))(2 - 1);
        2 === 0 ? 1 : 2 * (n => n === 0 ? 1 : n * factorial(n - 1))(2 - 1);
        false ? 1 : 2 * (n => n === 0 ? 1 : n * factorial(n - 1))(2 - 1);
        false ? 1 : 2 * (n => n === 0 ? 1 : n * factorial(n - 1))(2 - 1);
        2 * (n => n === 0 ? 1 : n * factorial(n - 1))(2 - 1);
        2 * (n => n === 0 ? 1 : n * factorial(n - 1))(2 - 1);
        2 * (n => n === 0 ? 1 : n * factorial(n - 1))(1);
        2 * (n => n === 0 ? 1 : n * factorial(n - 1))(1);
        2 * (1 === 0 ? 1 : 1 * (n => n === 0 ? 1 : n * factorial(n - 1))(1 - 1));
        2 * (1 === 0 ? 1 : 1 * (n => n === 0 ? 1 : n * factorial(n - 1))(1 - 1));
        2 * (false ? 1 : 1 * (n => n === 0 ? 1 : n * factorial(n - 1))(1 - 1));
        2 * (false ? 1 : 1 * (n => n === 0 ? 1 : n * factorial(n - 1))(1 - 1));
        2 * (1 * (n => n === 0 ? 1 : n * factorial(n - 1))(1 - 1));
        2 * (1 * (n => n === 0 ? 1 : n * factorial(n - 1))(1 - 1));
        2 * (1 * (n => n === 0 ? 1 : n * factorial(n - 1))(0));
        2 * (1 * (n => n === 0 ? 1 : n * factorial(n - 1))(0));
        2 * (1 * (0 === 0 ? 1 : 0 * (n => n === 0 ? 1 : n * factorial(n - 1))(0 - 1)));
        2 * (1 * (0 === 0 ? 1 : 0 * (n => n === 0 ? 1 : n * factorial(n - 1))(0 - 1)));
        2 * (1 * (true ? 1 : 0 * (n => n === 0 ? 1 : n * factorial(n - 1))(0 - 1)));
        2 * (1 * (true ? 1 : 0 * (n => n === 0 ? 1 : n * factorial(n - 1))(0 - 1)));
        2 * (1 * 1);
        2 * (1 * 1);
        2 * 1;
        2 * 1;
        2;
        2;"
      `)
    }),
    test('Even odd mutual', async () => {
      const code = `
      const odd = n => n === 0 ? false : even(n-1);
      const even = n => n === 0 || odd(n-1);
      even(1);
      `
      const steps = await codify(acornParser(code))
      expect(steps.join('\n')).toMatchInlineSnapshot(`
        "const odd = n => n === 0 ? false : even(n - 1);const even = n => n === 0 || odd(n - 1);even(1);
        const odd = n => n === 0 ? false : even(n - 1);const even = n => n === 0 || odd(n - 1);even(1);
        const even = n => n === 0 || (n => n === 0 ? false : even(n - 1))(n - 1);even(1);
        const even = n => n === 0 || (n => n === 0 ? false : even(n - 1))(n - 1);even(1);
        (n => n === 0 || (n => n === 0 ? false : even(n - 1))(n - 1))(1);
        (n => n === 0 || (n => n === 0 ? false : even(n - 1))(n - 1))(1);
        1 === 0 || (n => n === 0 ? false : (n => n === 0 || (n => n === 0 ? false : even(n - 1))(n - 1))(n - 1))(1 - 1);
        1 === 0 || (n => n === 0 ? false : (n => n === 0 || (n => n === 0 ? false : even(n - 1))(n - 1))(n - 1))(1 - 1);
        false || (n => n === 0 ? false : (n => n === 0 || (n => n === 0 ? false : even(n - 1))(n - 1))(n - 1))(1 - 1);
        false || (n => n === 0 ? false : (n => n === 0 || (n => n === 0 ? false : even(n - 1))(n - 1))(n - 1))(1 - 1);
        (n => n === 0 ? false : (n => n === 0 || (n => n === 0 ? false : even(n - 1))(n - 1))(n - 1))(1 - 1);
        (n => n === 0 ? false : (n => n === 0 || (n => n === 0 ? false : even(n - 1))(n - 1))(n - 1))(1 - 1);
        (n => n === 0 ? false : (n => n === 0 || (n => n === 0 ? false : even(n - 1))(n - 1))(n - 1))(0);
        (n => n === 0 ? false : (n => n === 0 || (n => n === 0 ? false : even(n - 1))(n - 1))(n - 1))(0);
        0 === 0 ? false : (n => n === 0 || (n => n === 0 ? false : even(n - 1))(n - 1))(0 - 1);
        0 === 0 ? false : (n => n === 0 || (n => n === 0 ? false : even(n - 1))(n - 1))(0 - 1);
        true ? false : (n => n === 0 || (n => n === 0 ? false : even(n - 1))(n - 1))(0 - 1);
        true ? false : (n => n === 0 || (n => n === 0 ? false : even(n - 1))(n - 1))(0 - 1);
        false;
        false;"
      `)
    })
  test('Mu term after substitution', async () => {
    const code = `
      const f = x => f;
      f(1);
    `
    const steps = await getSteps(convert(acornParser(code)))
    expect(steps.length).toBe(6)
    const lastStatement = ((steps[5].ast as StepperProgram).body[0] as StepperExpressionStatement)
      .expression
    expect(lastStatement.type).toBe('ArrowFunctionExpression')
    expect((lastStatement as StepperArrowFunctionExpression).name).toBe('f')
  })
})

describe('Alpha renaming', () => {
  test('Basic', async () => {
    const code = `
        const f = x => g();
        const g = () => x;
        const x = 1;
        f(0);
    `
    const steps = await codify(acornParser(code))
    expect(steps.join('\n')).toMatchInlineSnapshot(`
    "const f = x => g();const g = () => x;const x = 1;f(0);
    const f = x => g();const g = () => x;const x = 1;f(0);
    const g = () => x;const x = 1;(x => g())(0);
    const g = () => x;const x = 1;(x => g())(0);
    const x = 1;(x_1 => (() => x)())(0);
    const x = 1;(x_1 => (() => x)())(0);
    (x_1 => (() => 1)())(0);
    (x_1 => (() => 1)())(0);
    (() => 1)();
    (() => 1)();
    1;
    1;"
  `)
  }),
  test('Avoiding naming conflicts', async () => {
    const code = `
        const f = (x_1, x_3, x_2) => g();
        const g = () => x_1 + x_3 + x_2;
        const x_1 = 1;
        const x_3 = 3;
        const x_2 = 2;
        f(0);
    `
    const steps = await codify(acornParser(code))
    expect(steps.join('\n')).toMatchInlineSnapshot();
  })
})

describe('Function declaration', () => {
  test.todo("")
})

describe('SOURCE 0', () => {
  test.todo("")
})


describe('SOURCE 1', () => {
  test.todo("")
})

describe('SOURCE 2', () => {
  test.todo("")
})