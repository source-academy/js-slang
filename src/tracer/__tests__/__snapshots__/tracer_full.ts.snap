// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expressions No extra step for UnaryExpression 1`] = `"-(1 + 2);-(1 + 2);-3;-3;"`;
